src/Math/Quaternion.h

```diff
@@ -0,0 +1,216 @@
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+//  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+//  DEALINGS IN THE SOFTWARE.
+
+
+// COPIED NEARLY DIRECTLY FROM ADAFRUIT BNO055 LIBRARY
+#ifndef QUATERNION_H
+#define QUATERNION_H
+
+#include <math.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "Vector.h"
+
+namespace mmfs {
+
+class Quaternion {
+public:
+  Quaternion() : _w(1.0), _x(0.0), _y(0.0), _z(0.0) {}
+
+  Quaternion(double w, double x, double y, double z)
+      : _w(w), _x(x), _y(y), _z(z) {}
+
+  Quaternion(double w, Vector<3> vec)
+      : _w(w), _x(vec.x()), _y(vec.y()), _z(vec.z()) {}
+
+  double &w() { return _w; }
+  double &x() { return _x; }
+  double &y() { return _y; }
+  double &z() { return _z; }
+
+  double w() const { return _w; }
+  double x() const { return _x; }
+  double y() const { return _y; }
+  double z() const { return _z; }
+
+  double magnitude() const {
+    return sqrt(_w * _w + _x * _x + _y * _y + _z * _z);
+  }
+
+  void normalize() {
+    double mag = magnitude();
+    *this = this->scale(1 / mag);
+  }
+
+  Quaternion conjugate() const { return Quaternion(_w, -_x, -_y, -_z); }
+
+  void fromAxisAngle(const Vector<3> &axis, double theta) {
+    _w = cos(theta / 2);
+    // only need to calculate sine of half theta once
+    double sht = sin(theta / 2);
+    _x = axis.x() * sht;
+    _y = axis.y() * sht;
+    _z = axis.z() * sht;
+  }
+
+  // void fromMatrix(const Matrix<3> &m) {
+  //   double tr = m.trace();
+
+  //   double S;
+  //   if (tr > 0) {
+  //     S = sqrt(tr + 1.0) * 2;
+  //     _w = 0.25 * S;
+  //     _x = (m(2, 1) - m(1, 2)) / S;
+  //     _y = (m(0, 2) - m(2, 0)) / S;
+  //     _z = (m(1, 0) - m(0, 1)) / S;
+  //   } else if (m(0, 0) > m(1, 1) && m(0, 0) > m(2, 2)) {
+  //     S = sqrt(1.0 + m(0, 0) - m(1, 1) - m(2, 2)) * 2;
+  //     _w = (m(2, 1) - m(1, 2)) / S;
+  //     _x = 0.25 * S;
+  //     _y = (m(0, 1) + m(1, 0)) / S;
+  //     _z = (m(0, 2) + m(2, 0)) / S;
+  //   } else if (m(1, 1) > m(2, 2)) {
+  //     S = sqrt(1.0 + m(1, 1) - m(0, 0) - m(2, 2)) * 2;
+  //     _w = (m(0, 2) - m(2, 0)) / S;
+  //     _x = (m(0, 1) + m(1, 0)) / S;
+  //     _y = 0.25 * S;
+  //     _z = (m(1, 2) + m(2, 1)) / S;
+  //   } else {
+  //     S = sqrt(1.0 + m(2, 2) - m(0, 0) - m(1, 1)) * 2;
+  //     _w = (m(1, 0) - m(0, 1)) / S;
+  //     _x = (m(0, 2) + m(2, 0)) / S;
+  //     _y = (m(1, 2) + m(2, 1)) / S;
+  //     _z = 0.25 * S;
+  //   }
+  // }
+
```

Comment:
The commented code here should probably either be removed or uncommented.

Reply:
I'd like to keep it around in case we want to use it in the future. It's supposed to use the imu matrix class, not our own, which is why I commented it out. The rest of the class is verbatim Adafruit's Quaternion class and I didn't want to change it. The code should ideally not ever need to be modified, so leaving the comments there shouldn't be a big deal. I will add another comment explaining why it's commented.

Reply:
Ok, sounds good to me. Although if we don't end up using it for a while we should probably revisit removing just to keep the code cleaner.